libraries/AP_OSD/AP_OSD_ParamSetting.cpp

```diff
@@ -0,0 +1,418 @@
+        _param_incr.set_and_save(pincr);
+    } else {
+        guess_ranges(true);
+    }
+
+    _param_key.set_and_save(AP_Param::get_persistent_key(_current_token.key));
+    _param_idx.set_and_save(_current_token.idx);
+    _param_group.set_and_save(_current_token.group_element);
+    return true;
+}
+
+// guess the ranges and increment for the selected parameter
+void AP_OSD_ParamSetting::guess_ranges(bool force)
+{
+    if (_param->is_read_only()) {
+        return;
+    }
+
+    // check for statically configured setting metadata
+    if (set_from_metadata()) {
+        return;
+    }
+
+    // nothing statically configured so guess some appropriate values
+    float min = -1, max = 127, incr = 1;
+
+    if (_param != nullptr) {
+        switch (_param_type) {
+        case AP_PARAM_INT8:
+            break;
+        case AP_PARAM_INT16:
+        {
+            AP_Int16* p = (AP_Int16*)_param;
+            min = -1;
+            uint8_t digits = 0;
+            for (int16_t int16p = p->get(); int16p > 0; int16p /= 10) {
+                digits++;
+            }
+            incr = MAX(1, powf(10, digits - 2));
+            max = powf(10, digits + 1);
+            debug("Guessing range for value %d as %f -> %f, %f\n", p->get(), min, max, incr);
+        }
+            break;
+        case AP_PARAM_INT32:
+        {
+            AP_Int32* p = (AP_Int32*)_param;
+            min = -1;
+            uint8_t digits = 0;
+            for (int32_t int32p = p->get(); int32p > 0; int32p /= 10) {
+                digits++;
+            }
+            incr = MAX(1, powf(10, digits - 2));
+            max = powf(10, digits + 1);
+            debug("Guessing range for value %d as %f -> %f, %f\n", p->get(), min, max, incr);
+        }
+            break;
+        case AP_PARAM_FLOAT:
+        {
+            AP_Float* p = (AP_Float*)_param;
+
+            uint8_t digits = 0;
+            for (float floatp = p->get(); floatp > 1.0f; floatp /= 10) {
+                digits++;
+            }
+            float floatp = p->get();
+            if (digits < 1) {
+                if (!is_zero(floatp)) {
+                    incr = floatp / 100.0f; // move in 1% increments
+                } else {
+                    incr = 0.01f; // move in absolute 1% increments
+                }
+                max = 1.0;
+                min = 0.0f;
+            } else {
+                if (!is_zero(floatp)) {
+                    incr = floatp / 100.0f; // move in 1% increments
+                } else {
+                    incr = MAX(1, powf(10, digits - 2));
+                }
+                max = powf(10, digits + 1);
+                min = 0.0f;
+            }
+            debug("Guessing range for value %f as %f -> %f, %f\n", p->get(), min, max, incr);
+        }
+            break;
+        case AP_PARAM_VECTOR3F:
+        case AP_PARAM_NONE:
+        case AP_PARAM_GROUP:
+            break;
+        }
+
+        if (!_param_min.configured() || force) {
```

Comment:
configured() is expensive, need to make sure not being called at high rate

Reply:
only called when the user has made a change